packages/tools/src/Tool.ts

```diff
@@ -0,0 +1,147 @@
+/**
+ * A unique string identifier for each type of supported monorepo tool.
+ */
+export type ToolType = "bolt" | "lerna" | "pnpm" | "root" | "rush" | "yarn";
+
+/**
+ * A package.json access type.
+ */
+export type PackageAccessType = "public" | "restricted";
+
+/**
+ * An in-memory representation of a package.json file.
+ */
+export type PackageJSON = {
+  name: string;
+  version: string;
+  dependencies?: {
+    [key: string]: string;
+  };
+  peerDependencies?: {
+    [key: string]: string;
+  };
+  devDependencies?: {
+    [key: string]: string;
+  };
+  optionalDependencies?: {
+    [key: string]: string;
+  };
+  private?: boolean;
+  publishConfig?: {
+    access?: PackageAccessType;
+    directory?: string;
+    registry?: string;
+  };
+  manypkg?: {
+    [key: string]: string;
+  };
+};
+
+/**
+ * An individual package json structure, along with the directory it lives in,
+ * relative to the root of the current monorepo.
+ */
+export type Package = {
+  /**
+   * The pre-loaded package json structure.
+   */
+  packageJson: PackageJSON;
+
+  /**
+   * Absolute path to the directory containing this package.
+   */
+  dir: string;
+
+  /**
+   * Relative path to the directory containing this package, relative to the monorepo
+   * root (for a "root package", this is the string ".").
+   */
+  relativeDir: string;
+};
+
+/**
+ * A collection of packages, along with the monorepo tool used to load them,
+ * and (if supported by the tool) the associated "root" package.
+ */
+export type Packages = {
+  /**
+   * The underlying tool implementation for this monorepo.
+   */
+  tool: Tool;
+
+  /**
+   * A collection of disocvered packages.
+   */
+  packages: Package[];
+
+  /**
+   * If supported by the tool, this is the "root package" for the monorepo.
+   */
+  rootPackage?: Package;
+
+  /**
+   * The absolute path of the root directory of this monorepo.
+   */
+  rootDir: string;
+};
+
+/**
+ * An object representing the root of a specific monorepo, with the root
+ * directory and associated monorepo tool.
+ *
+ * Note that this type is currently not used by Tool definitions directly,
+ * but it is the suggested way to pass around a reference to a monorepo root
+ * directory and associated tool.
+ */
+export type MonorepoRoot = {
+  /**
+   * The absolute path to the root directory of this monorepo.
+   */
+  rootDir: string;
+
+  /**
+   * The underlying tool implementation for this monorepo.
+   */
+  tool: Tool;
+};
+
+/**
+ * Monorepo tools may throw this error if a caller attempts to get the package
+ * collection from a directory that is not a valid monorepo root.
+ */
+export class InvalidMonorepoError extends Error {}
```

Comment:
Two things here.

We are still using the old Babel config that transpiles a lot. Subclassing builtins has always been a problem when transpiled to ES5. Since this whole PR is a breaking change anyway - it's a good time to bump our syntax level, so we should tweak the config here:
https://github.com/Thinkmill/manypkg/blob/45b396d8c940db8d7ad0b9d6f9dd8dc695a15312/babel.config.js#L2

This can be configured like [here](https://github.com/changesets/changesets/blob/8447c677a37d1d23858059e0ce2b5cd98f4c24ae/babel.config.js#L3-L8) but we can already use `node: 14`.

The same information should be added to `package.json#engines`

A second thing is the question - should we export this publicly? I would probably prefer to avoid this because:
- this is a runtime export - so if we export this then we no longer get this: `import * as tools from '@manypkg/tools'`
- in more complex scenarios, node_modules layout etc relying on `instanceof` *externally* is a footgun anyway - so perhaps we shouldn't make it easy for people.

Reply:
I'm not dead-set on exporting it publicly, but I'll push back a little:

> this is a runtime export - so if we export this then we no longer get this: import * as tools from '@manypkg/tools'

I'm not sure what you mean here -- if you mean that `tools` is a collection of Tools, that is already not true, because `import *` also includes all our tool utilities like `Tool`, `PackageJSON`, `Packages`, etc. I think `@manypkg/tools` is a toolbox, and contains a collection of pre-baked Tools and all the ingredients you need to roll your own, and this Error is one of those ingredients.

> in more complex scenarios, node_modules layout etc relying on instanceof externally is a footgun anyway - so perhaps we shouldn't make it easy for people.

I suppose this is true although I do `instanceof` on Errors exported by e.g. AWS SDK without issue. If we wanted to truly fix this then I'd recommend adding a `static isInvalidMonorepoError(error: unknown)` boolean onto the Error, and adding a property `code: "InvalidMonorepoError"` to the class to perform the check. In this case I'd still argue it's useful to export because now you can use `if (InvalidMonorepoError.isInvalidMonorepoError(error))` in a try catch block.

Reply:
> I'm not sure what you mean here -- if you mean that tools is a collection of Tools, that is already not true, because import * also includes all our tool utilities like Tool, PackageJSON, Packages, etc.

All of those are just types - not runtime values. 
```ts
Object.keys(require('@manypkg/tools')) // ['BoltTool', 'LernaTool', 'PnpmTool', 'RootTool', 'RushTool', 'YarnTool', 'InvalidMonorepoError']
```

So `InvalidMonorepoError` is the only one at odds here.

> I suppose this is true although I do instanceof on Errors exported by e.g. AWS SDK without issue.

This is definitely an edge case that doesn't bite us too often but when it does... it hurts 😅 

> If we wanted to truly fix this then I'd recommend adding a static isInvalidMonorepoError(error: unknown) boolean onto the Error, and adding a property code: "InvalidMonorepoError" to the class to perform the check.

Yeah, that would work - at the expense of leaving that export in `@manypkg/tools`. Perhaps it's not worth trying to optimize the `import * as tools from '@manypkg/tools'` use case. I just thought that it would be kinda cool if we'd have that "trait".